pydrive2/fs.py

```diff
@@ -0,0 +1,476 @@
+import errno
+import io
+import logging
+import os
+import posixpath
+import threading
+from collections import defaultdict
+
+from fsspec.spec import AbstractFileSystem
+from funcy import cached_property, retry, wrap_prop, wrap_with
+from funcy.py3 import cat
+from tqdm.utils import CallbackIOWrapper
+
+from pydrive2.drive import GoogleDrive
+from pydrive2.fs_utils import IterStream
+
+logger = logging.getLogger(__name__)
+
+FOLDER_MIME_TYPE = "application/vnd.google-apps.folder"
+
+
+def _gdrive_retry(func):
+    def should_retry(exc):
+        from pydrive2.files import ApiRequestError
+
+        if not isinstance(exc, ApiRequestError):
+            return False
+
+        error_code = exc.error.get("code", 0)
+        result = False
+        if 500 <= error_code < 600:
+            result = True
+
+        if error_code == 403:
+            result = exc.GetField("reason") in [
+                "userRateLimitExceeded",
+                "rateLimitExceeded",
+            ]
+        if result:
+            logger.debug(f"Retrying GDrive API call, error: {exc}.")
+
+        return result
+
+    # 16 tries, start at 0.5s, multiply by golden ratio, cap at 20s
+    return retry(
+        16,
+        timeout=lambda a: min(0.5 * 1.618 ** a, 20),
+        filter_errors=should_retry,
+    )(func)
+
+
+class GDriveFileSystem(AbstractFileSystem):
+    def __init__(self, path, google_auth, trash_only=True, **kwargs):
+        self.path = path
+        self.root, self.base = self.split_path(self.path)
+        self.client = GoogleDrive(google_auth)
+        self._trash_only = trash_only
+        super().__init__(**kwargs)
+
+    def split_path(self, path):
+        parts = path.replace("//", "/").rstrip("/").split("/", 1)
+        if len(parts) == 2:
+            return parts
+        else:
+            return parts[0], ""
+
+    @wrap_prop(threading.RLock())
+    @cached_property
+    def _ids_cache(self):
+        cache = {
+            "dirs": defaultdict(list),
+            "ids": {},
+            "root_id": self._get_item_id(
+                self.path,
+                use_cache=False,
+                hint="Confirm the directory exists and you can access it.",
+            ),
+        }
+
+        self._cache_path_id(self.base, cache["root_id"], cache=cache)
+
+        for item in self._gdrive_list(
+            "'{}' in parents and trashed=false".format(cache["root_id"])
+        ):
+            item_path = posixpath.join(self.base, item["title"])
+            self._cache_path_id(item_path, item["id"], cache=cache)
+
+        return cache
+
+    def _cache_path_id(self, path, *item_ids, cache=None):
```

Comment:
it looks like we are accessing it from a few places now (and probably before), need to double check that we lock it properly everywhere

Reply:
Is there anything that might create a race condition here? The only possible suspect is `.append()` call, which we could probably guard by `if item_id not in cache["dirs"][path]:` to be safe.

Reply:
yep, my bad ... I see that create dir is protected separately. That should be enough probably. I was thinking about us creating duplicate directories (GDrive unfortunately allows that).